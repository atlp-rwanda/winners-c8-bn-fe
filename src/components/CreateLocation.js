import React, { Component } from 'react';
import { render } from 'react-dom';
import { connect } from 'react-redux';
import { toast, ToastContainer } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import locationIcon from '../../public/images/icons/locationIcon.svg';
import provinceIcon from '../../public/images/icons/province.svg';
import cityIcon from '../../public/images/icons/city.svg';

class CreateLocation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      location: {
        country: '',
        province: '',
        city: '',
      },
    };
  }
  handleSubmit = async (e) => {
    e.preventDefault();
    let myHeaders = new Headers();
    myHeaders.append('Content-Type', 'application/json');
    myHeaders.append('Authorization', `Bearer ${this.props.token}`);

    const { country, city, province } = this.state.location;
    let raw = JSON.stringify({ country, city, province });
    let requestOptions = {
      method: 'POST',
      headers: myHeaders,
      body: raw,
      redirect: 'follow',
    };
    this.setState({ wait: true });
    toast('Creating location . . .', {
       position: toast.POSITION.TOP_CENTER,
       });
    let result = await fetch(
      'https://winners-c8-bn-be-staging.herokuapp.com/api/locations',
      {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.props.token}`,
        },
        method: 'POST',
        body: JSON.stringify({ country, province, city }),
      }
    ).then((response) => response.json());

    this.setState({ responseMessage: result.message });
    toast.dismiss();

    if (result.status == 200) {
      toast.success('	The Location is successfully created.', {
        position: toast.POSITION.TOP_CENTER,
      });
    } else {
      toast.error(result.message || result.error, {
        position: toast.POSITION.TOP_CENTER,
      });
    }
  };
  render() {
    const { location } = this.state;
    return (
      <div className="white-bg">
        <div className="formContainer">
          <ToastContainer />
          <form className="flex-item">
            <h3 class="mb-32">CREATE LOCATION</h3>
            <div>
              <label htmlFor="country">Country</label>
              <div className="input-wrapper">
                <div className="icon">
                  <img src={locationIcon} />
                </div>
                <input
                  value={location.country}
                  onChange={(e) =>
                    this.setState({
                      ...this.state,
                      location: { ...location, country: e.target.value },
                    })
                  }
                  type="text"
                  id="country"
                  name="country"
                  className="form-control-accomodation"
                  data-testid="create-location-country"
                />
              </div>
            </div>
            <div>
              <label htmlFor="province">Province</label>
              <div className="input-wrapper">
                <div className="icon">
                  <img src={provinceIcon} />
                </div>
                <input
                  value={location.province}
                  onChange={(e) =>
                    this.setState({
                      ...this.state,
                      location: { ...location, province: e.target.value },
                    })
                  }
                  type="text"
                  id="province"
                  name="province"
                  className="form-control-accomodation"
                  data-testid="create-location-province"
                />
              </div>
            </div>
            <div>
              <label htmlFor="city">City</label>
              <div className="input-wrapper">
                <div className="icon">
                  <img src={cityIcon} />
                </div>
                <input
                  value={location.city}
                  onChange={(e) =>
                    this.setState({
                      ...this.state,
                      location: { ...location, city: e.target.value },
                    })
                  }
                  type="text"
                  id="city"
                  name="city"
                  className="form-control-accomodation"
                  data-testid="create-location-city"
                />
              </div>
            </div>
            {/* <div className="buttons-container"> */}
            <div className="buttons-wrapper">
              <button
                className="save_location "
                value="save"
                onClick={this.handleSubmit}
                data-testid="save-location"
              >
                Save
              </button>
            </div>
            {/* </div> */}
          </form>
        </div>
      </div>
    );
  }
}
const mapStateToProps = (state) => ({
  token: state.auth.token,
});
export default connect(mapStateToProps)(CreateLocation);
