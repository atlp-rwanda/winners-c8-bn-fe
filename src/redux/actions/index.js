// this folder will contain redux's actions