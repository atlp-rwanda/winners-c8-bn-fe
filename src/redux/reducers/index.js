// this folder will contain redux's reducers
import {accommodationsReducer, detailsAccommodationReducer, 
  AccommodationUpdateReducer,deleteAccommodationReducer,
  viewAccommodationReducer
}from "./accommodationReducers"
import { combineReducers } from 'redux';
import authReducer from './authReducer';
import navbarReducer from './navbarReducer';
import { registration } from './userReducers';
import { alert } from './alertReducers';
import {requestComments, postRequestCommentReducer, deleteRequestCommentReducer} from './requestReducer';
import requests from './requestReducer'
import resetReducer from './resetReducer';
import recoverReducer from './recoverReducer';
import userReducer from './SocialReducers';
import usersReducer from "./UserReducer";
import chatReducers from './chatReducers';
import { rolesReducer, assignRoleReducer } from "./RoleReducer";
import requestsReducer from './requestReducer'
import {
  fetchUserProfileReducer,
  updateUserProfileReducer,
} from './userProfileReducer';
import { statsReducer, chartStatsReducer } from './tripstatsReducers';
import PopularDestinationsReducer from './popularDestinationsReducer';

import { logoutReducer } from "./logoutReducer";

export default combineReducers({
  auth: authReducer,
  navbar: navbarReducer,
  requests,
  requestComments,
  postComment: postRequestCommentReducer,
  deleteComment: deleteRequestCommentReducer,
  userProfile: fetchUserProfileReducer,
  updateUserProfile: updateUserProfileReducer,
  registration,
  alert,
  chat: chatReducers,
  reset: resetReducer,
  recover: recoverReducer,
  user: userReducer,
  accommodations:accommodationsReducer,
  updatingAccomodation:AccommodationUpdateReducer,
  deletingAccommodation:deleteAccommodationReducer,
  viewAccommodation: viewAccommodationReducer,
  users: usersReducer,
  roles: rolesReducer,
  assignRole: assignRoleReducer,
  logout: logoutReducer,
  tripStats: statsReducer,
  chartTripStats: chartStatsReducer,
  PopularDestinations: PopularDestinationsReducer,
  requests: requestsReducer,
});


