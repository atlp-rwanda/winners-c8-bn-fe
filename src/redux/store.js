// this file will configure the redux's store
import { createStore } from "redux";
import rootReducer from "./reducers";

let authToken = ""; 
if(window.localStorage.getItem("auth-token")) {
  authToken = window.localStorage.getItem("auth-token");
}

const initialState = {
    auth: {token: authToken},
    navbar: {currentPage: "HOME"},
};
const store = createStore(
    rootReducer,
    initialState // sets initial state
  );

export default store;
