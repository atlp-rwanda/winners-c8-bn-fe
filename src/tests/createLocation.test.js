//accomodation tests
import React from 'react';
import {
  render,
  screen,
  cleanup,
  fireEvent,
  act,
  waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import {
  BrowserRouter as Router,
  MemoryRouter as MemoryRouter,
} from 'react-router-dom';
import thunk from 'redux-thunk';
import CreateLocation from '../components/CreateLocation';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { createRenderer } from 'react-dom/test-utils';

// import rootReducer from '../redux/reducers';
let store;
const initialState = {
  auth: { token: '' },
  location: {
    requestSent: false,
    responseData: {
      isSuccess: undefined,
      message: undefined,
    },
  },
};
beforeEach(() => {
  store = configureMockStore([thunk])(initialState);
});

afterEach(() => {
  cleanup;
});

describe('Testing rendering createLocation component', () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          status: 201,
          success: true,
          message: 'Location is created successfully',
        }),
    })
  );
  it('should validate invalid city', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () =>
          Promise.resolve({
            status: 404,
            success: true,
            message: 'Location is created successfully',
          }),
      })
    );
    await act(() =>
      render(
        <Provider store={store}>
          <Router>
            <CreateLocation />
          </Router>
        </Provider>
      )
    );
    const countryInput = screen.getByRole('textbox', { name: 'Country' });
    await act(async () => {
      fireEvent.change(countryInput, { value: '5335' });
    });
    const saveButton = screen.getByRole('button', { name: 'Save' });
    await act(async () => {
      fireEvent.click(saveButton);
    });
  });
});
