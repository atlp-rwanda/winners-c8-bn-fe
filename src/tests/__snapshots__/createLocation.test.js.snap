// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateLocation /> matches snapshot 1`] = `
<div
  className="white-bg"
>
  <div
    className="formContainer"
  >
    <div
      className="Toastify"
    />
    <form
      className="flex-item"
    >
      <h3
        class="mb-32"
      >
        CREATE LOCATION
      </h3>
      <div>
        <label
          htmlFor="country"
        >
          Country
        </label>
        <div
          className="input-wrapper"
        >
          <div
            className="icon"
          >
            <img
              src=""
            />
          </div>
          <input
            className="form-control-accomodation"
            data-testid="create-location-country"
            id="country"
            name="country"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <label
          htmlFor="province"
        >
          Province
        </label>
        <div
          className="input-wrapper"
        >
          <div
            className="icon"
          >
            <img
              src=""
            />
          </div>
          <input
            className="form-control-accomodation"
            data-testid="create-location-province"
            id="province"
            name="province"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <label
          htmlFor="city"
        >
          City
        </label>
        <div
          className="input-wrapper"
        >
          <div
            className="icon"
          >
            <img
              src=""
            />
          </div>
          <input
            className="form-control-accomodation"
            data-testid="create-location-city"
            id="city"
            name="city"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="buttons-wrapper"
      >
        <button
          className="save_location "
          data-testid="save-location"
          onClick={[Function]}
          value="save"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
